### ISSUE TITLE
'no-unused-prop-types' breaks when using destructuring in jsx attributes

### ISSUE THREAD

**Issue by ViggoV** [yannickcr/eslint-plugin-react#2222]
I was scratching my head over why no-unused-prop-types didn't trigger at all for some of my components, and after some experimentation it turns out it happens whenever I use destructuring in jsx attributes:

```javascript
const DumbComponent = props => {
  const someObj = { value: 619 };
  return (
    <SomeOtherComp {...someObj} />
  );
};

DumbComponent.propTypes = {
  unused_prop: PropTypes.string // doesn't trigger
};
```

but

```javascript
const DumbComponent = props => {
  const someObj = { value: 619 };
  return (
    <SomeOtherComp value={someObj.value} />
  );
};

DumbComponent.propTypes = {
  unused_prop: PropTypes.string // triggers
};
```

**Comment by ljharb** [comment]
I think you mean object spread - "destructuring" is a different feature.

This is because we can't statically know what you're spreading in - so when you spread, we currently assume they're all used.

However, in this case, the props object isn't what's being spread, so I'd assume it would still trigger.

**Comment by ViggoV** [comment]
Ah yes, i meant spread. Thanks for the correction.

**Comment by zdavis** [comment]
Seeing the same thing here.

No error for react/prop-types:
```
import React, { PureComponent } from "react";
import PropTypes from "prop-types";

export default class TestCase extends PureComponent {
  static propTypes = {};

  render() {
    const foo = { bar: this.props.bar };
    return <div {...foo} />;
  }
}
```
Error for react/prop-types is triggered in this case:
```
import React, { PureComponent } from "react";
import PropTypes from "prop-types";

export default class TestCase extends PureComponent {
  static propTypes = {};

  render() {
    const foo = { bar: this.props.bar };
    return <div />;
  }
}
```

**Comment by manneredboor** [comment]
I have faced same issue using typescript

Here it works correctly:
```jsx
type Props = {
  unusedPropType: string; // Triggered as supposed to be
  link?: string;
  onClick?: React.MouseEventHandler<HTMLDivElement | HTMLAnchorElement>;
  children?: React.ReactNode;
};

export function Tag({ link, onClick, children }: Props) {
  if (link) {
    return <Link to={link} onClick={onClick} children={children} />;
  }

  return <div onClick={onClick} children={children} />;
}
```

Here it is not:
```jsx

type Props = {
  unusedPropType: string; // Not triggered but supposed to be
  link?: string;
  onClick?: React.MouseEventHandler<HTMLDivElement | HTMLAnchorElement>;
  children?: React.ReactNode;
};

export function Tag({ link, onClick, children }: Props) {
  const props = {
    onClick,
    children,
  };

  if (link) {
    return <Link to={link} {...props} />;
  }

  return <div {...props} />;
}
```

![image](https://user-images.githubusercontent.com/7289992/110131102-da13b800-7dda-11eb-8db6-fc2abca15dac.png)
![image](https://user-images.githubusercontent.com/7289992/110131075-d1bb7d00-7dda-11eb-8bf8-dac02641925a.png)

**Comment by jwarby** [comment]
Also seeing the same issue when spreading a subset of props onto an element. It makes sense that it wouldn't trigger the rule if it is the `props` object itself being spread, but I think the rule should be able to figure out what is actually being used in the scenario described by @ViggoV and in my code sample below?

```javascript
const MyComponent = ({ foo }) => (
  <SomeOtherComponent {...{ foo }} />
);

MyComponent.propTypes = {
  foo: PropTypes.bool.isRequired,
  bar: PropTypes.bool.isRequired
};
```

From looking at the source, it seems like an `ignoreUnusedPropTypesValidation` flag will be set on any component that uses spread anywhere within, if I am not mistaken.
